homeassistant/components/zwave_js/api.py

```diff
@@ -1024,3 +1040,131 @@ async def get(self, request: web.Request, config_entry_id: str) -> web.Response:
+@async_get_node
+async def websocket_abort_firmware_update(
+    hass: HomeAssistant,
+    connection: ActiveConnection,
+    msg: dict,
+    node: Node,
+) -> None:
+    """Abort a firmware update."""
+    await node.async_abort_firmware_update()
+    connection.send_result(msg[ID])
+
+
+@websocket_api.require_admin
+@websocket_api.websocket_command(
+    {
+        vol.Required(TYPE): "zwave_js/subscribe_firmware_update_status",
+        vol.Required(ENTRY_ID): str,
+        vol.Required(NODE_ID): int,
+    }
+)
+@websocket_api.async_response
+@async_get_node
+async def websocket_subscribe_firmware_update_status(
+    hass: HomeAssistant,
+    connection: ActiveConnection,
+    msg: dict,
+    node: Node,
+) -> None:
+    """Subsribe to the status of a firmware update."""
+
+    @callback
+    def async_cleanup() -> None:
+        """Remove signal listeners."""
+        for unsub in unsubs:
+            unsub()
+
+    @callback
+    def forward_progress(event: dict) -> None:
+        progress: FirmwareUpdateProgress = event["firmware_update_progress"]
+        connection.send_message(
+            websocket_api.event_message(
+                msg[ID],
+                {
+                    "event": event["event"],
+                    "sent_fragments": progress.sent_fragments,
+                    "total_fragments": progress.total_fragments,
+                },
+            )
+        )
+
+    @callback
+    def forward_finished(event: dict) -> None:
+        finished: FirmwareUpdateFinished = event["firmware_update_finished"]
+        connection.send_message(
+            websocket_api.event_message(
+                msg[ID],
+                {
+                    "event": event["event"],
+                    "status": finished.status,
+                    "wait_time": finished.wait_time,
+                },
+            )
+        )
+
+    unsubs = [
+        node.on("firmware update progress", forward_progress),
+        node.on("firmware update finished", forward_finished),
+    ]
+    connection.subscriptions[msg["id"]] = async_cleanup
+
+    connection.send_result(msg[ID])
+
+
+class FirmwareUploadView(HomeAssistantView):
+    """View to upload firmware."""
+
+    url = r"/api/zwave_js/firmware/upload/{config_entry_id}/{node_id:\d+}"
+    name = "api:zwave_js:firmware:upload"
+
+    async def post(
+        self, request: web.Request, config_entry_id: str, node_id: str
+    ) -> web.Response:
+        """Handle upload."""
+        if not request["hass_user"].is_admin:
+            raise Unauthorized()
+        hass = request.app["hass"]
+        if config_entry_id not in hass.data[DOMAIN]:
+            raise web_exceptions.HTTPBadRequest
+
+        entry = hass.config_entries.async_get_entry(config_entry_id)
+        client = hass.data[DOMAIN][config_entry_id][DATA_CLIENT]
+        node = client.driver.controller.nodes.get(int(node_id))
+        if not node:
+            raise web_exceptions.HTTPBadRequest
```

Comment:
This should be a 404. 
```suggestion
            raise web_exceptions.HTTPNotFound
```